File: dart/algorithms/other/second_biggest_number.dart
import 'other.dart';

int secondBiggestNumber(List<int> input, {bool repeated = true}) {
  if (input.length < 2) return MIN_VALUE;

  var biggest = MIN_VALUE;
  var secondBiggest = MIN_VALUE;

  bool _isSecondBiggest(int input, int biggest, int secondBiggest) => !repeated
      ? input != biggest && input > secondBiggest
      : input > secondBiggest;

  for (var i = 0; i < input.length - 1; i++) {
    if (input[i] > biggest) {
      secondBiggest = biggest;
      biggest = input[i];
    } else if (_isSecondBiggest(input[i], biggest, secondBiggest)) {
      secondBiggest = input[i];
    }
  }

  return secondBiggest;
}

File: kotlin/algorithms/search/SecondBiggestNumber.kt
package algorithms.search

fun secondBiggestNumber(input: List<Int>, repeated: Boolean = true): Int {
    if (input.size < 2) return Int.MIN_VALUE

    var biggest = Int.MIN_VALUE
    var secondBiggest = Int.MIN_VALUE

    fun isSecondBiggest(input: Int, biggest: Int, secondBiggest: Int) =
        if (!repeated)
            input != biggest && input > secondBiggest
        else
            input > secondBiggest

    for (i in 0 until input.size) {
        if (input[i] > biggest) {
            secondBiggest = biggest
            biggest = input[i]
        } else if (isSecondBiggest(input[i], biggest, secondBiggest)) {
            secondBiggest = input[i]
        }
    }

    return secondBiggest
}